zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+        }
+
+        // Double-check that all the block lists are empty
+        #[cfg(any(dev, test))]
+        {
+            let qrange = self.queued.range(current_range).rev();
+            for (_, qblocks) in qrange {
+                debug_assert_eq!(qblocks.len(), 0);
+            }
+        }
+
+        // Now remove the empty vectors at all those heights
+        self.drop_range(current_range);
+
+        // Finally, update the checkpoint bounds
+        // The heights have to be valid here, because this part of the chain is valid
+        self.update_current_checkpoint_height(next_checkpoint_height);
+    }
+
+    /// Drop the queued blocks in `range`, and return an error for their
+    /// futures based on `err_str`.
+    ///
+    /// Also clears the corresponding keys in `self.queued`.
+    fn reject_range_with_error<R>(&mut self, range: R, err_str: &str)
+    where
+        R: RangeBounds<BlockHeight> + Clone,
+    {
+        let qrange = self.queued.range_mut(range.clone());
+        for (_, qblocks) in qrange {
+            for qblock in qblocks.drain(..) {
+                // Sending can fail, but there's nothing we can do about it.
+                let _ = qblock.tx.send(Err(err_str.to_string().into()));
+            }
+        }
+        self.drop_range(range);
+
+        // Do not update the current range.
+        // Instead, let the caller decide what to do.
+    }
+
+    /// Drop the queued blocks in `range`.
+    ///
+    /// Does not reject their futures, use `reject_range_with_error()` for that.
+    fn drop_range<R>(&mut self, range: R)
+    where
+        R: RangeBounds<BlockHeight>,
+    {
+        let drop_keys: Vec<BlockHeight> =
+            self.queued.range(range).map(|(key, _value)| *key).collect();
+        for k in drop_keys {
+            self.queued.remove(&k);
+        }
+        // Do not update the current range.
+        // Instead, let the caller decide what to do.
+    }
+}
+
+/// CheckpointVerifier rejects pending futures on drop.
+impl Drop for CheckpointVerifier {
+    fn drop(&mut self) {
+        self.reject_range_with_error(.., "checkpoint verifier was dropped");
+    }
+}
 
 /// The CheckpointVerifier service implementation.
 ///
-/// After verification, blocks are added to the underlying state service.
-impl<S> Service<Arc<Block>> for CheckpointVerifier<S>
-where
-    S: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>
-        + Send
-        + Clone
-        + 'static,
-    S::Future: Send + 'static,
-{
-    type Response = BlockHeaderHash;
-    type Error = Error;
+/// After verification, the block futures resolve to their hashes.
+impl Service<Arc<Block>> for CheckpointVerifier {
+    /// The CheckpointVerifier service has two layers of `Result`s, an inner
+    /// `checkpoint::Error`, and an outer `tokio::sync::oneshot::error::RecvError`.
+    type Response = Result<BlockHeaderHash, Error>;
+    type Error = oneshot::error::RecvError;
```

Comment:
It feels like this is an extra layer of complexity that it would be preferable to not put onto our callers.  The only case that a oneshot future resolves to an error is one where the sender is dropped before sending a message to the receiver.  But rather than asking the caller to handle this, I think that this is something that we can uphold as an invariant of the CheckpointService (e.g., by sending errors to all pending futures in the `Drop` impl).

Rather than returning a boxed receiver, we'd then return something like
```
async move {
    rx.await.expect("checkpointverifier does not leave dangling receivers")
}.boxed()
```

Reply:
> Indeed, this invariant is already maintained by the current code.

It seemed like a good idea, I had some tests hang when we left dangling receivers.

Reply:
> Rather than returning a boxed receiver, we'd then return something like
> 
> ```
> async move {
>     rx.await.expect("checkpointverifier does not leave dangling receivers")
> }.boxed()
> ```

Thanks, I was annoyed by the layers of results, and this was the pattern I needed to fix them.

Reply:
Changed to unwrap the RecvError result in 9d800c27469eef2adc6b88f9500874960bcb71d5